pipeline-scripts/release.groovy

```diff
@@ -460,4 +460,235 @@ def signArtifacts(Map signingParams) {
+                            /**
+                             * We need find the most recent version of 4.minor+1 and determine which
+                             * architectures it supports. If it only supports x86_64, the OTA team
+                             * only wants us to PR for 4.y+1 with a CPU suffix which limits upgrades
+                             * to 4.y+1 from this release's x86_64 image.
+                             */
+                            def spacedVersions = upgradeChannelVersions.join(' ')
+                            // Use sort -V to handle semver sorting of versions and choose the last
+                            def lastNextVer = sh(returnStdout: true, script: "echo ${spacedVersions} | tr ' ' '\\n' | grep ${major}[.]${minorNext}[.] | sort -V | tail -n 1").trim()
+                            if ( lastNextVer ) {
+
+                                /**
+                                 * @param name - The name of the release to look for
+                                 * @param arch - The architecture you want to know if the release is defined for
+                                 * @return Returns true if the release specified has support for the specified architecture
+                                 */
+                                def existsForArch = { name, arch ->
+                                    def suffix = getArchSuffix(arch)
+                                    // If there is an imagestream in the arch's api.ci namespace, it was built for that arch by ART.
+                                    // Check for both is/release and is/release-<ach> (the latter being an older style of release naming)
+                                    if ( sh(returnStdout: true, script: "oc get --kubeconfig ${buildlib.ciKubeconfig} --ignore-not-found -n ocp${suffix} -o=name is/${name} is/${name}${suffix}").trim() ) {
+                                        return true
+                                    } else {
+                                        return false
+                                    }
+                                }
+
+                                /**
+                                 * For each of the arches in the current release, see if a counterpart exists for the
+                                 * latest 4.y+1 release. If it is found, record that we are going to add an arch specific
+                                 * entry for the various channels (e.g. ["4.y.z+amd64", "4.y.z+s390x"] if both builds have
+                                 * x86_64 & s390x, but only 4.y.z has ppc64le)
+                                 */
+                                currentReleaseArches = buildlib.branch_arches('openshift-${major}.${minor}', true)
+                                for ( String arch : currentReleaseArches ) {
+                                    if ( existsForArch(lastNextVer, arch) ) {
```

Comment:
i'm wondering if the existence of the next release's imagestream is a sufficient condition to begin adding upgrade-ables in its channel. if i'm understanding correctly, a 4.4 RC release right now would generate a PR adding it into the 4.5 channels as an upgrade-able; the 4.5 channels would have nothing but 4.4 releases for months until we have a 4.5 RC, which would surely look confusing. i would think we don't want to add to upgrade channels until there's something there to upgrade *to*. we could either check the release-controller `4-stable` stream for the presence of something or the contents of the channel to see if there is anything for 4.5 before we start adding 4.4 entries. WDYT?

Reply:
There are two branches which try to prevent early creation / addition of 4.y to the 4.y+1 stream in that scenario. 
1. The 4.y+1 channel file must already exist: https://github.com/openshift/aos-cd-jobs/blob/a1ddb31b999258389035c5b7744fb33375df3302/pipeline-scripts/release.groovy#L526 . It will not be created by a 4.y release. 
2. If the file does exist, it must contain a 4.y+1 entry before the code will consider adding a 4.y entry: https://github.com/openshift/aos-cd-jobs/blob/a1ddb31b999258389035c5b7744fb33375df3302/pipeline-scripts/release.groovy#L540 . In your scenario, even if the 4.5 files exist, if they don't have a 4.5 entry, then `lastNextVer` in the code should be an empty string and skip setting `releasesForUpgradeChannel = true`.

Reply:
Thanks, I did not see the second part. Seems like you have it covered.